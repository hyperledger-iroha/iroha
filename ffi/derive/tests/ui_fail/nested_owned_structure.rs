use std::mem::MaybeUninit;

use iroha_ffi::{ffi_export, slice::OutBoxedSlice, FfiType};

type Nested = Vec<Vec<FfiStruct>>;

/// Ffi structure
#[derive(Clone, FfiType)]
pub struct FfiStruct;

/// Return nested structure
#[ffi_export]
pub fn return_nested() -> Nested {
    vec![vec![FfiStruct, FfiStruct], vec![FfiStruct, FfiStruct]]
}

fn main() {
    let mut params_len = MaybeUninit::uninit();
    let nested = OutBoxedSlice::from_uninit_slice(None, &mut params_len);
    unsafe { __return_nested(nested) };
}
