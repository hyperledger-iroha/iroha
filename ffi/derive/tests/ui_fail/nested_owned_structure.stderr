error[E0277]: the trait bound `IrVec<FfiStruct, Opaque<FfiStruct>>: NonLocal` is not satisfied
  --> tests/ui_fail/nested_owned_structure.rs:12:1
   |
12 | #[ffi_export]
   | ^^^^^^^^^^^^^ the trait `NonLocal` is not implemented for `IrVec<FfiStruct, Opaque<FfiStruct>>`
   |
   = help: the trait `NonLocal` is implemented for `IrVec<T, Transparent<T>>`
   = note: required because of the requirements on the impl of `COutPtr` for `IrVec<Vec<FfiStruct>, IrVec<FfiStruct, Opaque<FfiStruct>>>`
   = note: this error originates in the attribute macro `ffi_export` (in Nightly builds, run with -Z macro-backtrace for more info)
